### PR TITLE
[DependencyInjection] Simplified code in AutowirePass

### PR DESCRIPTION
**Pull request by hason**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -148,45 +148,17 @@ private function populateAvailableType($id, Definition $definition)
             $this->types[$type] = $id;
         }
 
-        // Cannot use reflection if the class isn't set
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Can you keep this comment?

Comment (hason) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: This code is a duplicate of https://github.com/hason/symfony/blob/autowire/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php#L231-L233

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 2]
```diff
@@ -148,45 +148,17 @@ private function populateAvailableType($id, Definition $definition)
...
-     *
-     * @param string           $id
-     * @param \ReflectionClass $reflectionClass
-     */
-    private function extractInterfaces($id, \ReflectionClass $reflectionClass)
-    {
-        foreach ($reflectionClass->getInterfaces() as $interfaceName => $reflectionInterface) {
-            $this->set($interfaceName, $id);
-
-            $this->extractInterfaces($id, $reflectionInterface);
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Interfaces ancestors are not extracted anymore?

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Move the comment then :)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: AFAIK, this is actually useless, as ``$reflectionClass->getInterfaces()`` gives you all implemented interfaces, even when they come from some inheritance. It does not give you only the interfaces appearing in the source code explicitly (but a check of the testsuite to ensure this case is covered should be necessary, to be sure that the code handles it right)

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 1 above)
Comment (hason) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Moved.

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php (hunk 2 above)
Comment (hason) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: Tests are enhanced.

**Comment by dunglas** [comment]
:+1:

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php [hunk 3]
```diff
@@ -317,13 +318,21 @@ public function __construct(A $a)
...
-class F implements EInterface
+interface IInterface
+{
+}
+
+class I implements IInterface
+{
+}
+
+class F extends I implements EInterface
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/AutowirePassTest.php]: you also needs the case of an interface extending another interface (and the class implementing the child one), to be sure it works fine too

**Comment by fabpot** [comment]
Thank you @hason.

**Pull request merged**